token-lending/program/src/processor.rs

```diff
@@ -1539,6 +1543,129 @@ fn process_repay_obligation_liquidity(
     Ok(())
 }
 
+#[inline(never)] // avoid stack frame limit
+fn process_oracleless_repay_obligation_liquidity(
+    program_id: &Pubkey,
+    liquidity_amount: u64,
+    accounts: &[AccountInfo],
+) -> ProgramResult {
+    if liquidity_amount == 0 {
+        msg!("Liquidity amount provided cannot be zero");
+        return Err(LendingError::InvalidAmount.into());
+    }
+    let account_info_iter = &mut accounts.iter();
+    let source_liquidity_info = next_account_info(account_info_iter)?;
+    let destination_liquidity_info = next_account_info(account_info_iter)?;
+    let repay_reserve_info = next_account_info(account_info_iter)?;
+    let obligation_info = next_account_info(account_info_iter)?;
+    let lending_market_info = next_account_info(account_info_iter)?;
+    let user_transfer_authority_info = next_account_info(account_info_iter)?;
+    let clock = &Clock::from_account_info(next_account_info(account_info_iter)?)?;
+    let token_program_id = next_account_info(account_info_iter)?;
+
+    let lending_market = LendingMarket::unpack(&lending_market_info.data.borrow())?;
+    if lending_market_info.owner != program_id {
+        msg!("Lending market provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &lending_market.token_program_id != token_program_id.key {
+        msg!("Lending market token program does not match the token program provided");
+        return Err(LendingError::InvalidTokenProgram.into());
+    }
+
+    let mut obligation = Obligation::unpack(&obligation_info.data.borrow())?;
+    if obligation_info.owner != program_id {
+        msg!("Obligation provided is not owned by the lending program");
+        return Err(LendingError::InvalidAccountOwner.into());
+    }
+    if &obligation.lending_market != lending_market_info.key {
+        msg!("Obligation lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+
+    for (index, liquidity) in obligation.borrows.iter_mut().enumerate() {
```

Comment:
you shouldn't have to refresh all reserves, just the one being touched by the repay

Reply:
/accumulate interest on all borrows for the obligation

Reply:
im pretty sure you do because otherwise the other ones wont accumulate properly on an actual refresh

Reply:
? wdym by that? their last updated slot/accumulated interest hasn't been updated so they would just be updated during the next full refresh

Reply:
hmm maybe i misread but in Obligation the staleness/last_updated_slot live at the top level so it felt safer to do this. though i do see that the cumulative_borrow_rate_wads lives on the obligationLiquidity so maybe it's ok hmm but would be odd because then different obligationLiquidities would be differently stale which currently never happens

Reply:
yea, don't really see any issues with having diff staleness on obligation liquidities because the entire obligation will be stale anyways

Reply:
im gonna read again to make sure lol. partly annoyed if that is true, but just because pr would have been much shorter :/